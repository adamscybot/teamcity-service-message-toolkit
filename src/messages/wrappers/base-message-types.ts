import {
  ApplyAttributeValueRules,
  TcBaseMessageInterface,
  TcBaseMultiAttributeMessageInterface,
  TcBaseSingleAttributeMessageInterface,
  TcMultiPropertyAttributeMap,
  TcMultiPropertyValidatedModifier,
} from '../types.js'
import {
  MessageValidationError,
  MessageValueAccessWhilstUnvalidated,
} from './errors.js'

export abstract class TcBaseMessage<ValidationModifier>
  implements TcBaseMessageInterface
{
  private _flowId: string | undefined = undefined
  private _validated: boolean = false

  constructor(flowId?: string) {
    this._flowId = flowId
    return this
  }

  flowId() {
    return this._flowId
  }

  messageName() {
    return (this.constructor as any)['messageName']
  }

  /**
   * Override this method with you own validation implementation. It is important
   * to ensure the concrete message return type is type narrowed to avoid any unuecessary checks
   * for `undefined` values in the consumer.
   *
   * @throws {MessageValidationError} when validation failed.
   */
  protected abstract doValidate(): this
  protected abstract doValidate<
    ValidatedModifier extends any
  >(): ValidatedModifier & this

  validate() {
    try {
      const validatedThis = this.doValidate<ValidationModifier>()
      this.setValidated(true)
      return validatedThis
    } catch (e) {
      this.setValidated(false)
      throw e
    }
  }

  protected setValidated(validated: boolean = true) {
    this._validated = validated
  }

  protected isValidated(): boolean {
    return this._validated
  }
}

export abstract class TcBaseSingleAttributeMessage<ValueType>
  extends TcBaseMessage<{ value: () => ValueType }>
  implements TcBaseSingleAttributeMessageInterface<ValueType>
{
  private _rawValue: string | undefined = undefined

  constructor(flowId?: string, rawValue?: string) {
    super(flowId)
    this._rawValue = rawValue
  }

  rawValue(value: string): this
  rawValue(): string | undefined
  rawValue(value?: string | undefined): string | undefined | this {
    if (value !== undefined) {
      this._rawValue = value
      return this
    }

    return this._rawValue
  }

  protected abstract getValue(): ValueType | undefined
  protected abstract setValue(value: ValueType): void

  value(value: ValueType): this
  value(): ValueType | undefined
  value(value?: ValueType | undefined): ValueType | undefined | this {
    if (value !== undefined) {
      this.setValue(value)
      return this
    }

    if (!this.isValidated()) throw new MessageValueAccessWhilstUnvalidated(this)
    return this.getValue()
  }
}

export class TcStringSingleAttributeMessage extends TcBaseSingleAttributeMessage<string> {
  protected override doValidate() {
    if (typeof this.value() !== 'string')
      throw new Error(
        `Value for message type '${this.messageName()}' is required but missing.`
      )

    return this
  }

  protected override getValue() {
    return this.rawValue()
  }

  protected override setValue(value: string) {
    this.rawValue(value)
    return this
  }
}

export abstract class TcBaseMultiAttributeMessage<
    AttributeMap extends TcMultiPropertyAttributeMap
  >
  extends TcBaseMessage<TcMultiPropertyValidatedModifier<AttributeMap>>
  implements TcBaseMultiAttributeMessageInterface<AttributeMap>
{
  protected _rawKwargs = {} as Record<string, string>

  constructor(
    rawKwargs: Record<string, string> = {} as Record<string, string>,
    flowId?: string
  ) {
    super(flowId)
    this._rawKwargs = rawKwargs
  }

  rawAttr(key: string, value: string): this
  rawAttr(key: string): string | undefined
  rawAttr(key: string, value?: string | undefined): string | undefined | this {
    if (value !== undefined) {
      this._rawKwargs[key] = value
      return this
    }

    return this._rawKwargs[key]
  }

  protected abstract getValue<Key extends keyof AttributeMap>(
    key: Key
  ): ApplyAttributeValueRules<AttributeMap, Key>
  protected abstract setValue<Key extends keyof AttributeMap>(
    key: Key,
    value: ApplyAttributeValueRules<AttributeMap, Key>
  ): void

  attr<Key extends keyof AttributeMap>(
    key: Key
  ): AttributeMap[Key]['type'] | undefined
  attr<Key extends keyof AttributeMap>(
    key: Key,
    value: AttributeMap[Key]['type']
  ): this
  attr<Key extends keyof AttributeMap>(
    key: Key,
    value?: AttributeMap[Key]['type']
  ): this | AttributeMap[Key]['type'] | undefined {
    if (value !== undefined) {
      this.setValue<Key>(key, value)
      return this
    }

    if (!this.isValidated())
      throw new MessageValueAccessWhilstUnvalidated(this, key as string)

    return this.getValue<Key>(key)
  }
}

class TcStringMultiAttributeMessage<
  StringAttributeMap extends Record<string, boolean>
> extends TcBaseMultiAttributeMessage<{
  [K in keyof StringAttributeMap]: {
    type: string
    required: StringAttributeMap[K]
  }
}> {
  protected override getValue<Key extends keyof StringAttributeMap>(
    key: Key
  ): ApplyAttributeValueRules<
    {
      [K in keyof StringAttributeMap]: {
        type: string
        required: StringAttributeMap[K]
      }
    },
    Key
  > {
    throw new Error('Method not implemented.')
  }
  protected override setValue<Key extends keyof StringAttributeMap>(
    key: Key,
    value: ApplyAttributeValueRules<
      {
        [K in keyof StringAttributeMap]: {
          type: string
          required: StringAttributeMap[K]
        }
      },
      Key
    >
  ): void {
    throw new Error('Method not implemented.')
  }

  protected override doValidate() {
    throw new Error('Method not implemented.')
  }
}
