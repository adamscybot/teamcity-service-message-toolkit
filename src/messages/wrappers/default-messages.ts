import {
  TcStringSingleAttributeMessage,
  TcStringMultiAttributeMessage,
} from './base-message-types.js'

/**
 * Used to indicate to the consumer the build number that represents this execution.
 */
export class TcBuildNumberMessage extends TcStringSingleAttributeMessage {
  public static readonly messageName = 'buildNumber'
}

export class TcPublishArtifacts extends TcStringSingleAttributeMessage {
  public static readonly messageName = 'publishArtifacts'
}

export class TcProgressMessage extends TcStringSingleAttributeMessage {
  public static readonly messageName = 'progressMessage'
}

export class TcProgressStartMessage extends TcStringSingleAttributeMessage {
  public static readonly messageName = 'progressStart'
}

export class TcProgressFinishMessage extends TcStringSingleAttributeMessage {
  public static readonly messageName = 'progressFinish'
}

export enum TcMessageStatus {
  NORMAL = 'NORMAL',
  WARNING = 'WARNING',
  FAILURE = 'FAILURE',
  ERROR = 'ERROR',
}

export class TcMessage extends TcStringMultiAttributeMessage<{
  text: { type: string; required: true }
  errorDetails: { type: string; required: false }
  status: { type: TcMessageStatus; required: false }
}> {
  public static readonly messageName = 'message'
}

export abstract class TcFlowMessage extends TcMultiAttributeMessage<'parent'> {
  parentId(parentFlowId: string) {
    this.attr('parent', parentFlowId)
    return this
  }
}

export class TcFlowStarted extends TcFlowMessage {
  public static readonly messageName = 'flowStarted'
}

export class TcFlowFinished extends TcFlowMessage {
  public static readonly messageName = 'flowFinished'
}

export class TcBlockOpened extends TcMultiAttributeMessage<
  'name' | 'description'
> {
  public static readonly messageName = 'blockOpened'

  name(name: string) {
    this.attr('name', name)
    return this
  }

  description(description: string) {
    this.attr('description', description)
    return this
  }
}

export class TcBlockClosed extends TcMultiAttributeMessage<'name'> {
  public static readonly messageName = 'blockClosed'

  name(name: string) {
    this.attr('name', name)
    return this
  }
}

export abstract class TcCompilationMessage extends TcMultiAttributeMessage<'compiler'> {
  compiler(compiler: string) {
    this.attr('compiler', compiler)
    return this
  }
}

export class TcCompilationStarted extends TcCompilationMessage {
  public static readonly messageName = 'compilationStarted'
}

export class TcCompilationFinished extends TcCompilationMessage {
  public static readonly messageName = 'compilationFinished'
}

export abstract class TcTestMessage<
  Keys extends string = ''
> extends TcMultiAttributeMessage<('name' | 'timestamp') | Keys> {
  /**
   * The unique name of this suite or test. This is used to capture the beginning/end of each
   * suite and test and so must not be duplicated in one execution.
   *
   * @param name The unique build number that represents this build
   * @example
   *
   * ```
   * message.name('Test com.mystuff.app')
   * ```
   */
  setName(name: string) {
    this.attr('name', name)
    return this
  }

  getName() {
    return this.attr('name')
  }

  /**
   * Allows a timestamp attribute to be added to this event. Useful if wanting to capture when
   * a certain test event happened.
   *
   * @param timestamp The time as a `Date` object that represents when this event happened.
   * @example
   *
   * ```
   * message.timestamp(Date.now())
   * ```
   */
  timestamp(timestamp: Date) {
    this.attr('timestamp', timestamp.toISOString())
    return this
  }
}

export class TcTestSuiteStarted extends TcTestMessage {
  public static readonly messageName = 'testSuiteStarted'
}

export class TcTestSuiteFinished extends TcTestMessage {
  public static readonly messageName = 'testSuiteFinished'
}

export class TcTestStarted extends TcTestMessage<'captureStandardOutput'> {
  public static readonly messageName = 'testStarted'

  /**
   * Consumers typically should assume that test runs use `testStdOut` and `testStdErr` messages to denote test output.
   * This can be set to true to avoid having to use those messages as if it is true, all output up to `testFinished` is considered
   * part of the test output.
   *
   * @param captureStandardOutput A boolean that if set to true, instructs the consumer to consider stdout up to when the `testFinished` fires as output for this test.
   * @example
   *
   * ```
   * message.captureStandardOutput(true)
   * ```
   */
  captureStandardOutput(captureStandardOutput: boolean) {
    this.attr('captureStandardOutput', String(captureStandardOutput))
    return this
  }
}

export class TcTestFinished extends TcTestMessage<'duration'> {
  public static readonly messageName = 'testFinished'

  duration(duration: number) {
    this.attr('timestamp', String(Math.round(duration)))
    return this
  }
}

export class TcTestIgnored extends TcTestMessage<'message'> {
  public static readonly messageName = 'testIgnored'

  message(message: string) {
    this.attr('message', message)
    return this
  }
}

export type TcTestFailedType = 'comparisonFailure'

export class TcTestFailed extends TcTestMessage<
  'message' | 'details' | 'expected' | 'actual' | 'type'
> {
  public static readonly messageName = 'testFailed'

  message(message: string) {
    this.attr('message', message)
    return this
  }

  details(details: string) {
    this.attr('details', details)
    return this
  }

  expected(expected: string) {
    this.attr('expected', expected)
    return this
  }

  actual(actual: string) {
    this.attr('actual', actual)
    return this
  }

  type(type: TcTestFailedType) {
    this.attr('type', type)
    return this
  }
}

export class TcTestStdOut extends TcTestMessage<'out'> {
  public static readonly messageName = 'testStdOut'

  out(out: string) {
    this.attr('out', out)
    return this
  }
}

export class TcTestStdErr extends TcTestMessage<'out'> {
  public static readonly messageName = 'testStdErr'

  out(out: string) {
    this.attr('out', out)
    return this
  }
}

export class TcTestRetrySupport extends TcMultiAttributeMessage<'enabled'> {
  public static readonly messageName = 'testRetrySupport'

  enabled(enabled: boolean) {
    this.attr('enabled', String(enabled))
    return this
  }
}

export const singleAttributeMessages = [
  TcBuildNumberMessage,
  TcPublishArtifacts,
  TcProgressMessage,
  TcProgressStartMessage,
  TcProgressFinishMessage,
] as const

export const multiAttributeMessages = [
  TcMessage,
  TcFlowStarted,
  TcFlowFinished,
  TcBlockOpened,
  TcBlockClosed,
  TcCompilationStarted,
  TcCompilationFinished,
  TcTestSuiteStarted,
  TcTestSuiteFinished,
  TcTestStarted,
  TcTestFinished,
  TcTestIgnored,
  TcTestFailed,
  TcTestStdOut,
  TcTestStdErr,
  TcTestRetrySupport,
] as const

const messages = [...singleAttributeMessages, ...multiAttributeMessages]
export default messages

export type MultiAttributeMessages = (typeof multiAttributeMessages)[number]
export type SingleAttributeMessages = (typeof singleAttributeMessages)[number]
export type Messages = (typeof messages)[number]
