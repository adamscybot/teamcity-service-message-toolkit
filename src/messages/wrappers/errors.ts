import { CustomError } from 'ts-custom-error'
import { TcBaseMessageInterface } from '../types.js'

export class MessageValueAccessWhilstUnvalidated extends CustomError {
  public constructor(message: TcBaseMessageInterface<any>, key?: string) {
    super(
      `Attemtped to access ${
        key === undefined ? '' : `'${key}' `
      } value on ${message} without first calling 'validate()'. The validate function must be called again if a setter was called. Call the validate function or use the other available methods to access the raw underlying values.`
    )
  }
}

export class MessageValidationError extends CustomError {
  public constructor(message: TcBaseMessageInterface<any>, reason: string) {
    super(`Message '${message}' failed validation. Reason: ${reason}`)
  }
}
