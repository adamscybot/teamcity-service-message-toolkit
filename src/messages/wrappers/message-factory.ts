import zod from 'zod'
import { ApplyAttributeValueRules, TcMultiPropertyAttributeMap } from '../types'
import { MessageValidationError } from './errors'
import { Key } from 'node:readline'

interface Message {
  flowId(): string | undefined
  messageName(): string
  validate(): boolean
  isValidated(): boolean
}

interface SingleAttributeMessage<ValueType> extends Message {
  rawValue(): string | undefined
  setValue(value: ValueType): void
  getValue(): ValueType | undefined
}

interface MultiAttributeMessage<
  AttributeMap extends TcMultiPropertyAttributeMap
> extends Message {
  rawAttr(key: string, value?: string | undefined): string | undefined
  setValue<Key extends keyof AttributeMap>(
    key: Key,
    value: ApplyAttributeValueRules<AttributeMap, Key>
  ): void
  getValue<Key extends keyof AttributeMap>(
    key: Key
  ): ApplyAttributeValueRules<AttributeMap, Key>
}

function createMessage({
  messageName,
  flowId,
  validate,
}: {
  readonly messageName: string
  readonly flowId?: string
  validate: () => boolean
}): Message {
  let _validated = false

  const message = {
    flowId: () => flowId,
    messageName: () => messageName,
    validate: () => {
      if (validate?.()) {
        _validated = true
      } else {
        _validated = false
        throw new MessageValidationError(message, ``)
      }

      return _validated
    },
    isValidated: () => _validated,
  }

  return message
}
// Factory function for SingleAttributeMessage, extending Message
export function createSingleAttributeMessage<ValueType>(
  messageName: string,
  rawValue?: string,
  flowId?: string,
  validate: () => boolean,
  getValue?: () => ValueType,
  setValue?: (value: ValueType) => void
): SingleAttributeMessage<ValueType> {
  let message = createMessage({ messageName, flowId, validate })

  return {
    ...message,
    rawValue: () => rawValue,
    setValue: setValue || (() => {}),
    getValue: getValue || (() => undefined),
  }
}

type AttributeAccessorMethods<T> = {
  fromString: (value: string) => T
  toString: (value: T) => string
}

type NonStringKeys<T> = {
  [K in keyof T]: T[K] extends string ? never : K
}[keyof T]

type OptionalKeys<T, Keys extends string> = Exclude<
  keyof T | Keys,
  NonStringKeys<T>
>

type MandatoryAttributeAccessMethods<T> = {
  [K in NonStringKeys<T>]: AttributeAccessorMethods<T[K]>
}

type OptionalAttributeAccessMethods<T, Keys extends string> = {
  [K in OptionalKeys<T, Keys>]?: AttributeAccessorMethods<string>
}

type AttributeAccessor<
  Keys extends string,
  AttributeValues extends Partial<Record<Keys, any>> = Partial<
    Record<Keys, string>
  >
> = MandatoryAttributeAccessMethods<AttributeValues> &
  OptionalAttributeAccessMethods<AttributeValues, Keys>

type AttributeAccessorOpt<
  Keys extends string,
  AttributeTypes extends Partial<Record<Keys, any>>
> = { accessors?: AttributeAccessor<Keys, AttributeTypes> }

type AttributeAccessorOpts<
  Keys extends string,
  AttributeTypes extends Partial<Record<Keys, any>>
> = AttributeTypes extends Partial<Record<Keys, string>>
  ? AttributeAccessorOpt<Keys, AttributeTypes>
  : Required<AttributeAccessorOpt<Keys, AttributeTypes>>

type MultiAttributeMessageOpts<
  Keys extends string,
  AttributeTypes extends Partial<Record<Keys, any>> = Partial<
    Record<Keys, string>
  >
> = {
  messageName: string
  rawKwargs: Record<string, string>
  flowId?: string
  validate: () => boolean
} & AttributeAccessorOpts<Keys, AttributeTypes>

export function createMultiAttributeMessage<
  Keys extends string,
  AttributeTypes extends Partial<Record<Keys, any>> = {}
>({
  messageName,
  rawKwargs,
  flowId,
  accessors,
  validate,
}: MultiAttributeMessageOpts<Keys, AttributeTypes>): any {
  let message = createMessage({ messageName, flowId, validate })
  let _rawKwargs = rawKwargs

  return {
    ...message,
    rawAttr: (key: string, value?: string) => {
      if (value !== undefined) {
        _rawKwargs[key] = value
      }

      return _rawKwargs[key]
    },

    attr<Key extends keyof AttributeMap>(
      key: Key,
      value?: AttributeMap[Key]['type']
    ): this | AttributeMap[Key]['type'] | undefined {
      if (value !== undefined) {
        if (setValue) {
          setValue(key, value)
        }
        return this
      }

      if (!this.isValidated())
        throw new MessageValueAccessWhilstUnvalidated(this, key as string)

      return getValue ? getValue(key) : undefined
    },
  }
}

createMultiAttributeMessage<'test' | 'test2' | 'number', { test: 2 }>({
  messageName: 'test',
  rawKwargs: {},
  validate() {
    return true
  },
})
