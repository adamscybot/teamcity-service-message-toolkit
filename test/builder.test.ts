import { describe, expect } from 'vitest'
import { z } from 'zod'

import { test } from './fixtures/wrappers'
import {
  TEST_MULTI_ATTR_MESSAGE_NAME,
  TEST_SINGLE_ATTR_MESSAGE_NAME,
} from './fixtures/messages'

describe('Message Builder', () => {
  describe('Single Attribute Messages', () => {
    test('Messages have static property defining their message type', ({
      singleAttrMessageBuilder,
    }) => {
      expect(singleAttrMessageBuilder.build().messageName).toBe(
        TEST_SINGLE_ATTR_MESSAGE_NAME
      )
    })

    test('Messages have message type property', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder.build()({})
      expect(message.messageName()).toBe(TEST_SINGLE_ATTR_MESSAGE_NAME)
    })

    test('Messages have flowId property', ({ singleAttrMessageBuilder }) => {
      const message = singleAttrMessageBuilder.build()({ flowId: '1234' })
      expect(message.flowId()).toBe('1234')
    })

    test('Messages expose raw value', ({ singleAttrMessageBuilder }) => {
      const message = singleAttrMessageBuilder.build()({ rawValue: 'test123' })
      expect(message.getRawValue()).toBe('test123')
    })

    test('Messages expose value as string by default', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder.build()({ rawValue: 'test123' })
      expect(message.getValue()).toBe('test123')
    })

    test('Messages dont allow value to be undefined by default', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder.build()({})
      expect(() => message.getValue()).toThrowErrorMatchingInlineSnapshot(`
        "[
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"string\\",
            \\"received\\": \\"undefined\\",
            \\"path\\": [],
            \\"message\\": \\"Required\\"
          }
        ]"
      `)
    })

    test('Message throws error when it violates schema', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder
        .schema(() => z.coerce.number())
        .build()({ rawValue: 'i am not a number' })

      const snapshot = `
        "[
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"number\\",
            \\"received\\": \\"nan\\",
            \\"path\\": [],
            \\"message\\": \\"Expected number, received nan\\"
          }
        ]"
      `
      expect(() => message.validate()).toThrowErrorMatchingSnapshot(snapshot)
      expect(() => message.getValue()).toThrowErrorMatchingSnapshot(snapshot)
    })

    test('Message transforms value in respect to schema', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder
        .schema(() => z.coerce.number())
        .build()({ rawValue: '999' })

      expect(() => message.validate()).not.toThrowError()
      expect(message.getValue()).toBe(999)
    })

    test('Messages with value can output as a TC service message string', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder.build()({
        rawValue: 'exampleAttr',
      })
      expect(message.toServiceMessageString()).toMatchInlineSnapshot(
        '"##teamcity[testSingleAttrMessage \'exampleAttr\']"'
      )
    })

    test('Messages without value can output as a TC service message string', ({
      singleAttrMessageBuilder,
    }) => {
      const messageWithValue = singleAttrMessageBuilder
        .schema((v) => v.default().optional())
        .build()({})
      expect(messageWithValue.toServiceMessageString()).toMatchInlineSnapshot(
        '"##teamcity[testSingleAttrMessage]"'
      )
    })

    test('Messages TC service string is correctly escaped', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder.build()({
        rawValue: "[Diff|cult\r\nmes'sage \u0134]",
      })
      expect(message.toServiceMessageString()).toMatchInlineSnapshot(
        "\"##teamcity[testSingleAttrMessage '|[Diff||cult|r|nmes|'sage |0x0134|]']\""
      )
    })

    test('Messages with flow ID gets flow id in service message string', ({
      singleAttrMessageBuilder,
    }) => {
      const message = singleAttrMessageBuilder.build()({
        rawValue: 'hello',
        flowId: '123',
      })
      expect(message.toServiceMessageString()).toMatchInlineSnapshot(
        "\"##teamcity[testSingleAttrMessage flowId='123' 'hello']\""
      )
    })
  })

  describe('Multi Attribute Messages', () => {
    test('Messages have static property defining their message type', ({
      multiAttrMessageBuilderWithAttrs,
    }) => {
      const factory = multiAttrMessageBuilderWithAttrs.build()
      expect(factory.messageName).toBe(TEST_MULTI_ATTR_MESSAGE_NAME)
    })

    test('Messages have message type property', ({
      multiAttrMessageBuilderWithAttrs,
    }) => {
      const message = multiAttrMessageBuilderWithAttrs.build()({
        rawKwargs: {},
      })
      expect(message.messageName()).toBe(TEST_MULTI_ATTR_MESSAGE_NAME)
    })

    test('Messages have flowId property', ({
      multiAttrMessageBuilderWithAttrs,
    }) => {
      const message = multiAttrMessageBuilderWithAttrs.build()({
        flowId: '1234',
        rawKwargs: { attr1: 'test', attr2: 'test', attr3: 'test' },
      })
      expect(message.flowId()).toBe('1234')
    })

    test('Messages expose raw attrs', ({
      multiAttrMessageBuilderWithAttrs,
    }) => {
      const attrs = { attr1: 'test1', attr2: 'test2', attr3: 'test3' } as const
      const message = multiAttrMessageBuilderWithAttrs.build()({
        rawKwargs: attrs,
      })
      expect(message.getRawAttr('attr1')).toBe('test1')
      expect(message.getRawAttr('attr2')).toBe('test2')
      expect(message.getRawAttr('attr3')).toBe('test3')
      expect(message.getRawAttrs()).toBe(attrs)
    })

    test('Messages expose attrs as string by default', ({
      multiAttrMessageBuilder,
    }) => {
      const attrs = { attr: 'stringExample' } as const
      const message = multiAttrMessageBuilder
        .schema((schema) => schema.attribute('attr').build())
        .build()({
        rawKwargs: attrs,
      })

      expect(message.getAttr('attr')).toBe('stringExample')
      expect(message.getAttrs()).toMatchInlineSnapshot(`
        {
          "attr": "stringExample",
        }
      `)
    })

    test('Messages dont allow value to be undefined by default', ({
      multiAttrMessageBuilder,
    }) => {
      const message = multiAttrMessageBuilder
        .schema((schema) => schema.attribute('attr').build())
        .build()({
        rawKwargs: {},
      })

      const snapshot = `
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"string\\",
          \\"received\\": \\"undefined\\",
          \\"path\\": [
            \\"attr\\"
          ],
          \\"message\\": \\"Required\\"
        }
      ]"
    `

      expect(() => message.getAttr('attr')).toThrowErrorMatchingSnapshot(
        snapshot
      )
      expect(() => message.getAttrs()).toThrowErrorMatchingSnapshot(snapshot)
    })

    test('Message throws error when it violates schema', ({
      multiAttrMessageBuilder,
    }) => {
      const message = multiAttrMessageBuilder
        .schema((schema) =>
          schema.attribute('example', () => z.coerce.number()).build()
        )
        .build()({ rawKwargs: { example: 'i am not a number' } })

      const snapshot = `
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"number\\",
          \\"received\\": \\"nan\\",
          \\"path\\": [
            \\"example\\"
          ],
          \\"message\\": \\"Expected number, received nan\\"
        }
      ]"
    `

      expect(() => message.validate()).toThrowErrorMatchingSnapshot(snapshot)
      expect(() => message.getAttr('example')).toThrowErrorMatchingSnapshot(
        snapshot
      )
      expect(() => message.getAttrs()).toThrowErrorMatchingSnapshot(snapshot)
    })

    test('Message transforms value in respect to schema', ({
      multiAttrMessageBuilder,
    }) => {
      const message = multiAttrMessageBuilder
        .schema((schema) =>
          schema.attribute('example', () => z.coerce.number()).build()
        )
        .build()({ rawKwargs: { example: '123' } })

      expect(() => message.validate()).not.toThrowError()
      expect(message.getAttr('example')).toBe(123)
    })

    test('Message is allowed to have extraneous properties as a raw value', ({
      multiAttrMessageBuilder,
    }) => {
      const message = multiAttrMessageBuilder
        .schema((schema) => schema.attribute('example').build())
        .build()({ rawKwargs: { example: '123', extraProperty: '123' } })

      expect(() => message.validate()).not.toThrowError()
      expect(message.getRawAttr('example')).toBe('123')
    })

    test('Messages can output as a TC service message string', ({
      multiAttrMessageBuilderWithAttrs,
    }) => {
      const message = multiAttrMessageBuilderWithAttrs.build()({
        rawKwargs: { attr1: 'test1', attr2: 'test2', attr3: 'test3' },
      })
      expect(message.toServiceMessageString()).toMatchInlineSnapshot(
        "\"##teamcity[testMultiAttrMessage attr1='test1' attr2='test2' attr3='test3']\""
      )
    })

    test('Messages TC service string is correctly escaped', ({
      multiAttrMessageBuilder,
    }) => {
      const message = multiAttrMessageBuilder
        .schema((schema) => schema.attribute('example').build())
        .build()({ rawKwargs: { example: "[Diff|cult\r\nmes'sage \u0134]" } })

      expect(message.toServiceMessageString()).toMatchInlineSnapshot(
        "\"##teamcity[testMultiAttrMessage example='|[Diff||cult|r|nmes|'sage |0x0134|]']\""
      )
    })

    test('Messages with flow ID gets flow id in service message string', ({
      multiAttrMessageBuilder,
    }) => {
      const message = multiAttrMessageBuilder
        .schema((schema) => schema.attribute('example').build())
        .build()({ rawKwargs: { example: 'test' }, flowId: '12345' })
      expect(message.toServiceMessageString()).toMatchInlineSnapshot(
        "\"##teamcity[testMultiAttrMessage flowId='12345' example='test']\""
      )
    })
  })
})
