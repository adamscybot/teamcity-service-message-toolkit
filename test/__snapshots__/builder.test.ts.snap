// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Message Builder > Multi Attribute Messages > Message throws error when it violates schema > 
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"number\\",
          \\"received\\": \\"nan\\",
          \\"path\\": [
            \\"example\\"
          ],
          \\"message\\": \\"Expected number, received nan\\"
        }
      ]"
     1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [
      \\"example\\"
    ],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Message Builder > Multi Attribute Messages > Message throws error when it violates schema > 
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"number\\",
          \\"received\\": \\"nan\\",
          \\"path\\": [
            \\"example\\"
          ],
          \\"message\\": \\"Expected number, received nan\\"
        }
      ]"
     2`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [
      \\"example\\"
    ],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Message Builder > Multi Attribute Messages > Message throws error when it violates schema > 
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"number\\",
          \\"received\\": \\"nan\\",
          \\"path\\": [
            \\"example\\"
          ],
          \\"message\\": \\"Expected number, received nan\\"
        }
      ]"
     3`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [
      \\"example\\"
    ],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Message Builder > Multi Attribute Messages > Messages dont allow value to be undefined by default > 
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"string\\",
          \\"received\\": \\"undefined\\",
          \\"path\\": [
            \\"attr\\"
          ],
          \\"message\\": \\"Required\\"
        }
      ]"
     1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"attr\\"
    ],
    \\"message\\": \\"Required\\"
  }
]"
`;

exports[`Message Builder > Multi Attribute Messages > Messages dont allow value to be undefined by default > 
      "[
        {
          \\"code\\": \\"invalid_type\\",
          \\"expected\\": \\"string\\",
          \\"received\\": \\"undefined\\",
          \\"path\\": [
            \\"attr\\"
          ],
          \\"message\\": \\"Required\\"
        }
      ]"
     2`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"undefined\\",
    \\"path\\": [
      \\"attr\\"
    ],
    \\"message\\": \\"Required\\"
  }
]"
`;

exports[`Message Builder > Single Attribute Messages > Message throws error when it violates schema > 
        "[
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"number\\",
            \\"received\\": \\"nan\\",
            \\"path\\": [],
            \\"message\\": \\"Expected number, received nan\\"
          }
        ]"
       1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;

exports[`Message Builder > Single Attribute Messages > Message throws error when it violates schema > 
        "[
          {
            \\"code\\": \\"invalid_type\\",
            \\"expected\\": \\"number\\",
            \\"received\\": \\"nan\\",
            \\"path\\": [],
            \\"message\\": \\"Expected number, received nan\\"
          }
        ]"
       2`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"number\\",
    \\"received\\": \\"nan\\",
    \\"path\\": [],
    \\"message\\": \\"Expected number, received nan\\"
  }
]"
`;
